File: JIWON/059 다리를 지나는 트럭.cpp
#include <string>
#include <vector>
#include <queue>

using namespace std;

int solution(int bridge_length, int weight, vector<int> truck_weights) {
    int answer = 0;
    queue<int> q;
    
    if(truck_weights.size() == 1){
        answer += bridge_length+1;
        
        return answer;
    }
    
    int bridge_weight = 0;
    int truckIdx = 0;
    
    while(true){
        answer++;
        
        if(q.size() == bridge_length){ // 다리가 다 찼으면
            bridge_weight -= q.front();
            q.pop();
        }
        
        if(bridge_weight + truck_weights[truckIdx] <= weight){ // 무게가 안넘으면
            bridge_weight += truck_weights[truckIdx];
            
            if(truckIdx == truck_weights.size() - 1) {
                answer += bridge_length;
                break;
            }
            
            q.push(truck_weights[truckIdx]);
            truckIdx++;
        } else{ // 무게가 넘으면
            q.push(0);
        }
    }
    
    return answer;
}

File: lesunny/059_다리를지나는트럭.java
import java.util.Queue;
import java.util.LinkedList;

class Solution {
    public int solution(int bridge_length, int weight, int[] truck_weights) {
        int answer = 0;
        Queue<Integer> bridge = new LinkedList<>();
        
        int current_weight = 0;
        for(int truck : truck_weights)
        {
            while(true){
                if(bridge.isEmpty()){
                    bridge.offer(truck);
                    answer++;
                    current_weight+=truck;
                    break;
                }else if(bridge.size() == bridge_length){
                    current_weight-=bridge.poll();
                }else if(current_weight + truck <= weight){
                    bridge.offer(truck);
                    answer++;
                    current_weight+=truck;
                    break;
                }else{
                    bridge.offer(0);
                    answer++;
                }
            }
        }
        
        return answer + bridge_length;
    }
}
